pkg/raft/term_cache_test.go

```diff
@@ -0,0 +1,505 @@
+				last:     entryID{term: 10, index: 18},
+				indexAndExpectedTerm: [][2]uint64{
+					{10, 0},
+					{11, 6},
+					{12, 7}, {13, 7}, {14, 7},
+					{15, 9}, {16, 9},
+					{17, 9}, {18, 10},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Initialize cache with size 4
+			cache := NewTermCache(tc.cacheSize, tc.initialID)
+
+			// First append to establish initial state
+			initialSlice := tc.initialID.append(tc.initTerms...)
+			cache.truncateAndAppend(initialSlice.LogSlice)
+
+			// Second append with new entries
+			appendSlice := tc.appendInitialID.append(tc.appendTerms...)
+			cache.truncateAndAppend(appendSlice.LogSlice)
+
+			// Verify final state
+			require.Equal(t, tc.want.cacheLen, len(cache.cache))
+			require.Equal(t, tc.want.first, cache.firstEntry())
+			require.Equal(t, tc.want.last, cache.lastEntry())
+
+			// Verify capacity hasn't increased
+			require.Equal(t, tc.cacheSize, uint64(cap(cache.cache)),
+				"capacity should remain at initial size")
+
+			checkTerm(t, &cache, tc.want.indexAndExpectedTerm)
+		})
+	}
+}
+
+func TestTermCacheInitialTruncateAndAppend(t *testing.T) {
+	testCases := []struct {
+		name      string
+		initialID entryID
+		terms     []uint64
+		want      struct {
+			cacheLen int
+			first    entryID
+			last     entryID
+		}
+	}{
+		{
+			name:      "evicts oldest entries when full",
+			initialID: entryID{term: 5, index: 10},
+			// append [t5/11, t6/12, t6/13, t7/14, t7/15, t8/16, t9/17]
+			terms: []uint64{5, 6, 6, 7, 7, 8, 9},
+			want: struct {
+				cacheLen int
+				first    entryID
+				last     entryID
+			}{
+				cacheLen: 4,                           // Maximum size
+				first:    entryID{term: 6, index: 12}, // Oldest entries evicted
+				last:     entryID{term: 9, index: 17}, // Last entry
+			},
+		},
+		{
+			name:      "appends small amounts of term flips, not full, no evict",
+			initialID: entryID{term: 5, index: 10},
+			// append [t6/11, t7/12]
+			terms: []uint64{6, 7},
+			want: struct {
+				cacheLen int
+				first    entryID
+				last     entryID
+			}{
+				cacheLen: 3,                           // We are not full yet
+				first:    entryID{term: 5, index: 10}, // Initial entry
+				last:     entryID{term: 7, index: 12}, // Last entry
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Initialize cache with size 4
+			cache := NewTermCache(4, tc.initialID)
+
+			// Append entries using the helper
+			slice := tc.initialID.append(tc.terms...)
+			cache.truncateAndAppend(slice.LogSlice)
+
+			// Verify results
+			require.Equal(t, tc.want.cacheLen, len(cache.cache))
+			require.Equal(t, tc.want.first, cache.firstEntry())
+			require.Equal(t, tc.want.last, cache.lastEntry())
+
+			// Verify capacity hasn't increased
+			require.Equal(t, 4, cap(cache.cache),
+				"capacity should remain at initial size")
+		})
+	}
+}
+
+func TestNeverEmpty(t *testing.T) {
```

Comment:
The checks in this test could be extended to "check all invariants":

- not empty
- `term/index` are sorted and increasing

And then we don't need a separate test for that. Instead, make every test do `tc.checkInvariants(t)` after every modifying operation on `tc`.